shell/platform/android/io/flutter/embedding/android/KeyEmbedderResponder.java

```diff
@@ -0,0 +1,365 @@
+
+package io.flutter.embedding.android;
+
+import android.view.KeyEvent;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import io.flutter.embedding.android.KeyboardMap.PressingGoal;
+import io.flutter.embedding.android.KeyboardMap.TogglingGoal;
+import io.flutter.plugin.common.BinaryMessenger;
+import java.util.HashMap;
+
+/**
+ * A {@link KeyboardManager.Responder} of {@link KeyboardManager} that handles events by sending
+ * processed information in {@link KeyData}.
+ *
+ * <p>This class corresponds to the HardwareKeyboard API in the framework.
+ */
+public class KeyEmbedderResponder implements KeyboardManager.Responder {
+  private static final String TAG = "KeyEmbedderResponder";
+
+  // Maps KeyEvent's action and repeatCount to a KeyData type.
+  private static KeyData.Type getEventType(KeyEvent event) {
+    final boolean isRepeatEvent = event.getRepeatCount() > 0;
+    switch (event.getAction()) {
+      case KeyEvent.ACTION_DOWN:
+        return isRepeatEvent ? KeyData.Type.kRepeat : KeyData.Type.kDown;
+      case KeyEvent.ACTION_UP:
+        return KeyData.Type.kUp;
+      default:
+        throw new AssertionError("Unexpected event type");
+    }
+  }
+
+  // The messenger used to send Flutter key events to the framework.
+  //
+  // On `handleEvent`, Flutter events are marshalled into byte buffers in the format specified by
+  // `KeyData.toBytes`.
+  @NonNull private final BinaryMessenger messenger;
+  // The keys being pressed currently, mapped from physical keys to logical keys.
+  @NonNull private final HashMap<Long, Long> pressingRecords = new HashMap<>();
+  // Map from logical key to toggling goals.
+  //
+  // Besides immutable configuration, the toggling goals are also used to store the current enabling
+  // states in their `enabled` field.
+  @NonNull private final HashMap<Long, TogglingGoal> togglingGoals = new HashMap<>();
+
+  @NonNull
+  private final KeyboardManager.CharacterCombiner characterCombiner =
+      new KeyboardManager.CharacterCombiner();
+
+  public KeyEmbedderResponder(BinaryMessenger messenger) {
+    this.messenger = messenger;
+    for (final TogglingGoal goal : KeyboardMap.getTogglingGoals()) {
+      togglingGoals.put(goal.logicalKey, goal);
+    }
+  }
+
+  // Get the physical key for this event.
+  //
+  // The returned value is never null.
+  private Long getPhysicalKey(@NonNull KeyEvent event) {
+    final Long byMapping = KeyboardMap.scanCodeToPhysical.get((long) event.getScanCode());
+    if (byMapping != null) {
+      return byMapping;
+    }
+    return KeyboardMap.kAndroidPlane + event.getScanCode();
+  }
+
+  // Get the logical key for this event.
+  //
+  // The returned value is never null.
+  private Long getLogicalKey(@NonNull KeyEvent event) {
+    final Long byMapping = KeyboardMap.keyCodeToLogical.get((long) event.getKeyCode());
+    if (byMapping != null) {
+      return byMapping;
+    }
+    return KeyboardMap.kAndroidPlane + event.getKeyCode();
```

Comment:
You might want to use `KeyboardMap.kAndroidPlane & event.getKeyCode()` instead: addition of a large value can cause it to overflow, and then it won't be in the Android plane anymore. Not that it is likely at all, just safer. (and other places in this PR). I mean, I guess in that case it would be messed up anyhow, but still.